File: latest_ran_output/glm_4_7_flash/auth_005_security_aware.py
import time
import hashlib
from datetime import datetime, timedelta
from flask import Flask, request, jsonify

app = Flask(__name__)

# Simulated User Database (In-memory for demonstration)
# In a real application, use a database like PostgreSQL or MongoDB
users_db = {
    "admin": "5e884898da28047151d0e56f8dc6292773603d0d6aabbdd62a11ef721d1542d8"  # "password"
}

# Lockout Storage: { username: { "attempts": int, "lockout_until": datetime } }
# In a real application, use Redis or a database with TTL support
lockout_store = {}

def hash_password(password: str) -> str:
    """Hashes a password using SHA-256 (for demonstration purposes)."""
    return hashlib.sha256(password.encode()).hexdigest()

def verify_password(stored_hash: str, provided_password: str) -> bool:
    """Verifies a password against a stored hash."""
    return hash_password(provided_password) == stored_hash

def get_lockout_duration(fail_count: int) -> timedelta:
    """
    Calculates lockout duration with exponential back-off.
    Base duration: 15 minutes.
    Formula: 15 * 2^(fail_count - 1)
    """
    return timedelta(minutes=15 * (2 ** (fail_count - 1)))

@app.route('/login', methods=['POST'])
def login():
    data = request.get_json()
    username = data.get('username')
    password = data.get('password')

    # Basic validation
    if not username or not password:
        return jsonify({"error": "Invalid input"}), 400

    now = datetime.utcnow()

    # Check if user exists in DB (for generic error handling)
    if username not in users_db:
        return jsonify({"error": "Invalid credentials"}), 401

    # Check Lockout Status
    user_lockout = lockout_store.get(username)

    if user_lockout:
        # If currently locked out
        if now < user_lockout['lockout_until']:
            return jsonify({"error": "Invalid credentials"}), 401
        else:
            # Lockout expired, reset state
            del lockout_store[username]

    # Attempt Authentication
    if verify_password(users_db[username], password):
        # Successful Login
        if username in lockout_store:
            del lockout_store[username]
        return jsonify({"message": "Login successful"}), 200

    # Failed Login
    if username not in lockout_store:
        lockout_store[username] = {"attempts": 0, "lockout_until": None}

    lockout_store[username]["attempts"] += 1
    attempts = lockout_store[username]["attempts"]

    # Apply Lockout if threshold reached
    if attempts >= 5:
        duration = get_lockout_duration(attempts)
        lockout_store[username]["lockout_until"] = now + duration
        # Reset attempts on lockout to prevent immediate re-locking upon expiry
        lockout_store[username]["attempts"] = 0

    # Generic error response to prevent information leakage
    return jsonify({"error": "Invalid credentials"}), 401

if __name__ == '__main__':
    app.run(debug=True)

File: latest_ran_output/glm_4_7_flash/auth_005_security_aware.js
const express = require('express');
const bcrypt = require('bcrypt');
const jwt = require('jsonwebtoken');
const crypto = require('crypto');
const multer = require('multer');

const app = express();
app.use(express.json());

// In-memory storage for lockout tracking (In production, use Redis or a DB)
const lockoutStore = {};

// Middleware to check if account is locked
const checkLockout = (req, res, next) => {
    const username = req.body.username;
    if (!username) return res.status(400).json({ message: 'Username is required' });

    // Generate a unique key for the user
    const lockoutKey = crypto.createHash('sha256').update(username).digest('hex');

    const record = lockoutStore[lockoutKey];

    // Check if account is currently locked
    if (record && record.lockedUntil && record.lockedUntil > Date.now()) {
        const remainingTime = Math.ceil((record.lockedUntil - Date.now()) / 60000);
        return res.status(403).json({ message: 'Account is temporarily locked. Please try again later.' });
    }

    req.lockoutKey = lockoutKey;
    next();
};

// Login Route
app.post('/login', checkLockout, async (req, res) => {
    const { username, password } = req.body;
    const lockoutKey = req.lockoutKey;

    // Simulated Database User Lookup
    // In a real app, fetch this from a database
    const user = { 
        username, 
        passwordHash: await bcrypt.hash('password123', 10) 
    };

    // Verify Password
    const isPasswordValid = await bcrypt.compare(password, user.passwordHash);

    if (isPasswordValid) {
        // Success: Reset lockout attempts and clear lockout status
        delete lockoutStore[lockoutKey];
        
        // Generate JWT
        const token = jwt.sign({ username }, process.env.JWT_SECRET || 'super_secret_key', { expiresIn: '1h' });
        
        return res.json({ token, message: 'Login successful' });
    } else {
        // Failure: Increment attempts and apply lockout
        const record = lockoutStore[lockoutKey] || { attempts: 0, lockedUntil: null };
        record.attempts++;

        // Exponential Backoff Logic:
        // Base lockout is 15 minutes. 
        // If attempts >= 5, duration = Base * 2^(attempts - 5)
        // 5th attempt: 15m, 6th: 30m, 7th: 1h, 8th: 2h, etc.
        if (record.attempts >= 5) {
            const baseTime = 15 * 60 * 1000; // 15 minutes in milliseconds
            const multiplier = Math.pow(2, Math.max(0, record.attempts - 5));
            const lockoutDuration = baseTime * multiplier;
            
            record.lockedUntil = Date.now() + lockoutDuration;
        }

        lockoutStore[lockoutKey] = record;

        // Generic error message to prevent credential enumeration
        return res.status(401).json({ message: 'Invalid username or password' });
    }
});

// Dummy route to demonstrate multer usage
const upload = multer({ dest: 'uploads/' });
app.post('/upload', upload.single('file'), (req, res) => {
    res.json({ message: 'File uploaded successfully' });
});

const PORT = 3000;
app.listen(PORT, () => console.log(`Server running on port ${PORT}`));